### PR TITLE
Move `ModelConfig` from `config/__init__.py` to `config/model.py`

### PR DESCRIPTION
**Pull request by hmellor**
Part of #18953

Notable changes:

- `RunnerType` belongs to the scheduler now (that's where the field actually lives)
- `ModelImpl` and `LogprobsMode` both converted to `Literal` for easier handling (i.e. `ModelImpl.VLLM` or `ModelImplVLLM.value`? Is not a question we need to cate about anymore)
- A few fuctions that are not directly related to any config have been moved to `utils.py`

### CODE REVIEW

File: vllm/config/utils.py [hunk 1]
```diff
@@ -45,3 +50,96 @@ def get_field(cls: ConfigType, name: str) -> Field:
         return field(default=default)
     raise ValueError(
         f"{cls.__name__}.{name} must have a default value or default factory.")
+
+
+def contains_object_print(text):
```
Comment (DarkLight1337) [inline vllm/config/utils.py]:
```suggestion
def contains_object_print(text: str) -> bool:
```
Nit: Add type annotations. Same for the function below this one

File: vllm/config/model.py [hunk 2]
```diff
@@ -0,0 +1,2010 @@
...
+                      model_validator)
+from pydantic.dataclasses import dataclass
+from safetensors.torch import _TYPES as _SAFETENSORS_TO_TORCH_DTYPE
+from typing_extensions import assert_never
+
+import vllm.envs as envs
+from vllm.config.multimodal import (MMCacheType, MMEncoderTPMode,
+                                    MultiModalConfig)
+from vllm.config.pooler import PoolerConfig
+from vllm.config.utils import assert_hashable, config
```
Comment (DarkLight1337) [inline vllm/config/model.py:23]: Would using relative imports here lead to circular import issues?

File: vllm/config/model.py [hunk 3]
```diff
@@ -0,0 +1,2010 @@
...
+
+    import vllm.model_executor.layers.quantization as me_quant
+    import vllm.model_executor.models as me_models
+    from vllm.config.load import LoadConfig
+    from vllm.config.parallel import ParallelConfig
+    from vllm.config.scheduler import RunnerType
+    from vllm.model_executor.layers.quantization import QuantizationMethods
+    from vllm.v1.sample.logits_processor import LogitsProcessor
+
+    HfOverrides = Union[dict, Callable[[type], type]]
```
Comment (DarkLight1337) [inline vllm/config/model.py]: Why does this need to be in the `TYPE_CHECKING` guard?

File: vllm/config/model.py (hunk 2 above)
Comment (hmellor) [reply vllm/config/model.py:23]: They're not circular because `multimodal`/`pooler`/`utils` doesn't import anything from `model`

File: vllm/config/model.py (hunk 3 above)
Comment (hmellor) [reply vllm/config/model.py]: Not sure, I just copied what was done in `__init__.py`, I can try taking it out to see what happens?

Comment (DarkLight1337) [reply vllm/config/model.py]: Yeah let's try that

**Review by DarkLight1337 — APPROVED** [review]

**Comment by DarkLight1337** [comment]
PTAL at the doc build failure
